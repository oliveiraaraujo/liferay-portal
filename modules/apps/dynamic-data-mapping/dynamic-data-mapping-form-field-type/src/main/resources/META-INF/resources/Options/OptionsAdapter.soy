{namespace OptionsAdapter}

/**
 * OptionsAdapter
 */
{template .render}
	<div key="ddm-field-type-options"></div>
{/template}