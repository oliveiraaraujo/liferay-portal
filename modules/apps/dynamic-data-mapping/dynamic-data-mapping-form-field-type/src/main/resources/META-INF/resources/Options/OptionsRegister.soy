{namespace OptionsRegister}

{deltemplate PageRenderer.RegisterFieldType variant="'options'"}
	{call OptionsAdapter.render data="all" /}
{/deltemplate}