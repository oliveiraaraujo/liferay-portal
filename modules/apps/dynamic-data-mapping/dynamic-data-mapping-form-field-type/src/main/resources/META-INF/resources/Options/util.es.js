/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {normalizeFieldName} from 'dynamic-data-mapping-form-renderer';

export const random = a => {
	return a
		? (a ^ ((Math.random() * 16) >> (a / 4))).toString(16)
		: ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, random);
};

export const compose = (...fns) =>
	fns.reduceRight((f, g) => (...xs) => {
		const r = g(...xs);

		return Array.isArray(r) ? f(...r) : f(r);
	});

export const isOptionValueGenerated = (
	defaultLanguageId,
	editingLanguageId,
	option
) => {
	if (defaultLanguageId !== editingLanguageId) {
		return false;
	}

	if (option.edited) {
		return false;
	}

	if (
		new RegExp(`^${Liferay.Language.get('option')}\\d*$`).test(option.value)
	) {
		return true;
	}

	if (
		new RegExp(`^${option.value.replace(/\d+$/, '')}\\d*`).test(
			normalizeFieldName(option.label)
		)
	) {
		return true;
	}

	return true;
};

/**
 * Deduplicates the value by checking if there is a
 * value in the fields, always incrementing an integer
 * in front of the value to be friendly for the user.
 */
export const dedupValue = (fields, value, id) => {
	let counter = 0;

	const recursive = (fields, currentValue) => {
		const field = fields.find(field => field.value === currentValue);

		if (field && field.id !== id) {
			counter += 1;
			recursive(fields, value + counter);
		}
		else {
			value = currentValue;
		}
	};

	recursive(fields, value);

	return value;
};

/**
 * O normalize value impede que value seja nulo ou undefined.
 * If the value is null or undefined, normalize follows a
 * verification order and the final stage of normalization
 * is to deduplicate the value if necessary.
 *
 * 1. If the current value is null, use the label
 * 2. If the current label is null, use the string Option
 */
export const normalizeValue = (fields, currentField) => {
	const {label, value: prevValue} = currentField;
	let value = prevValue ? prevValue : label;

	if (!value) {
		value = Liferay.Language.get('option');
	}

	value = dedupValue(fields, value, currentField.id);

	return normalizeFieldName(value);
};

export const normalizeFields = fields => {
	return fields.map((field, index) => {
		if (fields.length - 1 === index) {
			return field;
		}

		return {
			...field,
			value: normalizeValue(fields, field),
		};
	});
};
