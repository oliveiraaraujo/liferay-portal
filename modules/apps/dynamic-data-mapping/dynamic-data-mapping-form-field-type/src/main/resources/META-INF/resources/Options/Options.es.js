/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import './OptionsRegister.soy';

import ClayIcon from '@clayui/icon';
import classNames from 'classnames';
import React, {useMemo, useState} from 'react';
import {DndProvider} from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';

import {FieldBaseProxy} from '../FieldBase/ReactFieldBase.es';
import {Main as KeyValue} from '../KeyValue/KeyValue.es';
import getConnectedReactComponentAdapter from '../util/ReactComponentAdapter.es';
import {connectStore} from '../util/connectStore.es';
import DnD from './DnD.es';
import DragPreview from './DragPreview.es';
import templates from './OptionsAdapter.soy';
import {
	compose,
	dedupValue,
	isOptionValueGenerated,
	normalizeFields,
	random,
} from './util.es';

const defaultOption = {label: '', value: ''};

const Option = React.forwardRef(
	({children, className, disabled, onClick, showCloseButton, style}, ref) => (
		<div
			className={classNames('ddm-field-options', className)}
			style={style}
		>
			<span
				className={classNames('ddm-options-drag', {
					disabled,
				})}
				ref={disabled ? null : ref}
			>
				<ClayIcon symbol="drag" />
			</span>

			<div className="ddm-option-entry">
				{children}

				{showCloseButton && (
					<button
						className="close close-modal"
						onClick={onClick}
						type="button"
					>
						<ClayIcon symbol="times" />
					</button>
				)}
			</div>
		</div>
	)
);

const Options = ({
	children,
	defaultLanguageId,
	disabled,
	editingLanguageId,
	onChange,
	value = {},
}) => {
	const normalizedValue = useMemo(() => {
		const formattedValue = {...value};

		Object.keys(value).forEach(languageId => {
			if (defaultLanguageId !== languageId) {
				formattedValue[languageId] = formattedValue[languageId].filter(
					({value}) => !!value
				);
			}
		});

		return formattedValue;
	}, [defaultLanguageId, value]);

	const [fields, setFields] = useState(() => {
		const options =
			normalizedValue[editingLanguageId] ||
			normalizedValue[defaultLanguageId] ||
			[];

		return [
			...options.map(option => ({
				...option,
				generateKeyword: isOptionValueGenerated(
					defaultLanguageId,
					editingLanguageId,
					options,
					option
				),
				id: random(),
			})),
			defaultLanguageId === editingLanguageId
				? {...defaultOption, generateKeyword: true, id: random()}
				: false,
		].filter(Boolean);
	});

	const fieldsFilter = fields => {
		const _fields = [...fields];

		_fields.splice(_fields.length - 1, 1);

		return {...normalizedValue, [editingLanguageId]: _fields};
	};

	const clone = (...args) => {
		return [[...fields], ...args];
	};

	const dedup = (fields, index, property, value) => {
		const {generateKeyword, id} = fields[index];

		if (property === 'value' && generateKeyword) {
			value = dedupValue(
				fields,
				value ? value : Liferay.Language.get('option'),
				id
			);
		}

		return [fields, index, property, value];
	};

	const set = fields => {
		setFields(fields);
		onChange(fieldsFilter(fields));
	};

	const add = (fields, index, property, value) => {
		fields[index][property] = value;

		if (defaultLanguageId === editingLanguageId) {
			fields.push({
				...defaultOption,
				generateKeyword: true,
				id: random(),
			});
		}

		return [fields, index, property, value];
	};

	const change = (fields, index, property, value) => {
		const {edited, label} = fields[index];

		fields[index][property] = value;
		fields[index]['edited'] =
			edited || (value && value !== label && property === 'value');

		return [fields, index, property, value];
	};

	const normalize = fields => {
		return [normalizeFields(fields)];
	};

	const handleDelete = (fields, index) => {
		fields.splice(index, 1);

		return [fields];
	};

	const move = (fields, data) => {
		const {itemPosition, targetPosition} = data;

		if (itemPosition === fields.length - 1) {
			return [fields];
		}

		const item = {...fields[itemPosition]};
		const newTargetPosition =
			targetPosition > itemPosition ? targetPosition - 1 : targetPosition;

		fields.splice(itemPosition, 1);
		fields.splice(newTargetPosition, 0, item);

		return [fields];
	};

	const composedAdd = compose(clone, dedup, add, set);
	const composedBlur = compose(clone, normalize, set);
	const composedChange = compose(clone, dedup, change, set);
	const composedDelete = compose(clone, handleDelete, set);
	const composedMove = compose(clone, move, set);

	return (
		<div className="ddm-field-options-container">
			<DragPreview component={Option}>{children}</DragPreview>
			{fields.map((option, index) => (
				<DnD
					index={index}
					key={option.id}
					onDragEnd={composedMove}
					option={option}
				>
					<Option
						disabled={disabled}
						onClick={() => composedDelete(index)}
						showCloseButton={
							!(fields.length - 1 === index) && !disabled
						}
					>
						{children({
							handleBlur: composedBlur,
							handleField: !(fields.length - 1 === index)
								? composedChange.bind(this, index)
								: composedAdd.bind(this, index),
							index,
							option,
						})}
					</Option>
				</DnD>
			))}
		</div>
	);
};

const OptionsProxy = connectStore(
	({
		defaultLanguageId = themeDisplay.getLanguageId(),
		dispatch,
		editingLanguageId = themeDisplay.getLanguageId(),
		emit,
		placeholder = Liferay.Language.get('enter-an-option'),
		readOnly,
		required,
		store,
		value = {},
		visible,
		...otherProps
	}) => (
		<DndProvider backend={HTML5Backend}>
			<FieldBaseProxy
				{...otherProps}
				dispatch={dispatch}
				readOnly={readOnly}
				store={store}
				visible={visible}
			>
				<Options
					defaultLanguageId={defaultLanguageId}
					disabled={readOnly}
					editingLanguageId={editingLanguageId}
					onChange={value => emit('fieldEdited', {}, value)}
					value={value}
				>
					{({handleBlur, handleField, index, option}) => (
						<KeyValue
							dispatch={dispatch}
							generateKeyword={option.generateKeyword}
							keyword={option.value}
							keywordReadOnly={
								defaultLanguageId !== editingLanguageId
							}
							name={`option${index}`}
							onKeywordBlur={handleBlur}
							onKeywordChange={(event, value, generate) => {
								handleField('generateKeyword', generate);
								handleField('value', value);
							}}
							onTextBlur={handleBlur}
							onTextChange={event =>
								handleField('label', event.target.value)
							}
							placeholder={placeholder}
							readOnly={option.disabled}
							required={required}
							showLabel={false}
							store={store}
							value={option.label}
							visible={visible}
						/>
					)}
				</Options>
			</FieldBaseProxy>
		</DndProvider>
	)
);

const ReactOptionsAdapter = getConnectedReactComponentAdapter(
	OptionsProxy,
	templates
);

export {ReactOptionsAdapter};
export default ReactOptionsAdapter;
