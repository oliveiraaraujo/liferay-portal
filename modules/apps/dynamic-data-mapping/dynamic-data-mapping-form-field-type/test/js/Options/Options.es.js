/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {fireEvent} from '@testing-library/react';

import Options from '../../../src/main/resources/META-INF/resources/Options/Options.es';
import withContextMock from '../__mocks__/withContextMock.es';

let component;
const spritemap = 'icons.svg';

const OptionsWithContextMock = withContextMock(Options);

const optionsValue = {
	[themeDisplay.getLanguageId()]: [
		{
			label: 'Option 1',
			value: 'Option1',
		},
		{
			label: 'Option 2',
			value: 'Option2',
		},
	],
};

describe('Options', () => {
	beforeEach(() => jest.useFakeTimers());

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('shows the options', () => {
		component = new OptionsWithContextMock({
			name: 'options',
			spritemap,
			value: optionsValue,
		});

		expect(component).toMatchSnapshot();
	});

	it('shows an empty option when value is an array of size 1', () => {
		component = new OptionsWithContextMock({
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Option',
						value: 'Option',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInputs = element.querySelectorAll('.ddm-field-text');

		expect(labelInputs.length).toEqual(2);
		expect(labelInputs[0].value).toEqual('Option');
		expect(labelInputs[1].value).toEqual('');

		const valueInputs = element.querySelectorAll('.key-value-input');

		expect(valueInputs.length).toEqual(2);
		expect(valueInputs[0].value).toEqual('Option');
		expect(valueInputs[1].value).toEqual('');
	});

	it('does not show an empty option when translating', () => {
		component = new OptionsWithContextMock({
			defaultLanguageId: themeDisplay.getLanguageId(),
			editingLanguageId: 'pt_BR',
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Option',
						value: 'Option',
					},
				],
				pt_BR: [
					{
						label: 'Option',
						value: 'Option',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInputs = element.querySelectorAll('.ddm-field-text');

		expect(labelInputs.length).toEqual(1);
	});

	it('edits the value of an option based on the label', () => {
		component = new OptionsWithContextMock({
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Option',
						value: 'Option',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInputs = element.querySelectorAll('.ddm-field-text');

		fireEvent.change(labelInputs[0], {
			target: {
				value: 'Hello',
			},
		});

		jest.runAllTimers();

		const valueInputs = element.querySelectorAll('.key-value-input');

		expect(valueInputs[0].value).toEqual('Hello');
	});

	it('inserts a new empty option when editing the last option', () => {
		component = new OptionsWithContextMock({
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Option',
						value: 'Option',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInputs = element.querySelectorAll('.ddm-field-text');

		fireEvent.change(labelInputs[1], {
			target: {
				value: 'Hello',
			},
		});

		jest.runAllTimers();

		const valueInputs = element.querySelectorAll('.key-value-input');

		expect(valueInputs.length).toEqual(labelInputs.length + 1);
	});

	it('does not insert a new empty option automatically if translating', () => {
		component = new OptionsWithContextMock({
			defaultLanguageId: themeDisplay.getLanguageId(),
			editingLanguageId: 'pt_BR',
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Option',
						value: 'Option',
					},
				],
				pt_BR: [
					{
						label: 'Option',
						value: 'Option',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInputs = element.querySelectorAll('.ddm-field-text');

		fireEvent.input(labelInputs[0], {target: {value: 'Hello'}});

		jest.runAllTimers();

		const valueInputs = element.querySelectorAll('.key-value-input');

		expect(valueInputs.length).toEqual(labelInputs.length);
	});

	it('deduplication of value when adding a new option', () => {
		component = new OptionsWithContextMock({
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Foo',
						value: 'Foo',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInputs = element.querySelectorAll('.ddm-field-text');

		fireEvent.input(labelInputs[1], {target: {value: 'Foo'}});

		const valueInputs = element.querySelectorAll('.key-value-input');

		expect(valueInputs[1].value).toEqual('Foo1');
	});

	it('deduplication of the value when editing the value', () => {
		component = new OptionsWithContextMock({
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Bar',
						value: 'Bar',
					},
					{
						label: 'Foo',
						value: 'Foo',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInputs = element.querySelectorAll('.ddm-field-text');

		fireEvent.input(labelInputs[1], {target: {value: 'Bar'}});

		const valueInputs = element.querySelectorAll('.key-value-input');

		expect(valueInputs[1].value).toEqual('Bar1');
	});

	it('adds a value to the value property when the label is empty', () => {
		component = new OptionsWithContextMock({
			name: 'options',
			spritemap,
			value: {
				[themeDisplay.getLanguageId()]: [
					{
						label: 'Bar',
						value: 'Bar',
					},
				],
			},
		});

		jest.runAllTimers();

		const {element} = component;
		const labelInput = element.querySelector('.ddm-field-text');

		fireEvent.input(labelInput, {target: {value: ''}});

		const valueInput = element.querySelector('.key-value-input');

		expect(valueInput.value).toBe('option');
	});
});
