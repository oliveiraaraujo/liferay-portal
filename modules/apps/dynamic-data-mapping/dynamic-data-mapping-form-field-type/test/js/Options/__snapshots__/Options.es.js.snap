// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options shows the options 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="options">
    <div class="ddm-field-options-container">
      <div class="ddm-options-target"></div>
      <div class="ddm-field-options"><span class="ddm-options-drag" draggable="true"><svg class="lexicon-icon lexicon-icon-drag" role="presentation">
            <use xlink:href="icons.svg#drag"></use>
          </svg></span>
        <div class="ddm-option-entry">
          <div class="form-group" data-field-name="option0">
            <div class="form-group" data-field-name="keyValueLabeloption0"><input aria-label="text" name="keyValueLabeloption0" placeholder="enter-an-option" type="text" class="form-control ddm-field-text" value="Option 1"></div>
            <div class="active form-text key-value-editor"><label class="control-label key-value-label">field-name:</label><input class="key-value-input" type="text" value="Option1"></div>
          </div><button class="close close-modal" type="button"><svg class="lexicon-icon lexicon-icon-times" role="presentation">
              <use xlink:href="icons.svg#times"></use>
            </svg></button>
        </div>
      </div>
      <div class="ddm-options-target"></div>
      <div class="ddm-field-options"><span class="ddm-options-drag" draggable="true"><svg class="lexicon-icon lexicon-icon-drag" role="presentation">
            <use xlink:href="icons.svg#drag"></use>
          </svg></span>
        <div class="ddm-option-entry">
          <div class="form-group" data-field-name="option1">
            <div class="form-group" data-field-name="keyValueLabeloption1"><input aria-label="text" name="keyValueLabeloption1" placeholder="enter-an-option" type="text" class="form-control ddm-field-text" value="Option 2"></div>
            <div class="active form-text key-value-editor"><label class="control-label key-value-label">field-name:</label><input class="key-value-input" type="text" value="Option2"></div>
          </div><button class="close close-modal" type="button"><svg class="lexicon-icon lexicon-icon-times" role="presentation">
              <use xlink:href="icons.svg#times"></use>
            </svg></button>
        </div>
      </div>
      <div class="ddm-options-target"></div>
      <div class="ddm-field-options"><span class="ddm-options-drag" draggable="true"><svg class="lexicon-icon lexicon-icon-drag" role="presentation">
            <use xlink:href="icons.svg#drag"></use>
          </svg></span>
        <div class="ddm-option-entry">
          <div class="form-group" data-field-name="option2">
            <div class="form-group" data-field-name="keyValueLabeloption2"><input aria-label="text" name="keyValueLabeloption2" placeholder="enter-an-option" type="text" class="form-control ddm-field-text" value=""></div>
            <div class="active form-text key-value-editor"><label class="control-label key-value-label">field-name:</label><input class="key-value-input" type="text" value=""></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
